media_driver/linux/common/ddi/media_libva.cpp

```diff
@@ -6002,6 +6002,128 @@ DdiMedia_CopyInternal(
     return vaStatus;
 }
 
+//!
+//! \brief  media copy
+//!
+//! \param  [in] ctx
+//!         Pointer to VA driver context
+//! \param  [in] dst_obj
+//!         VA copy object dst.
+//! \param  [in] src_obj
+//!         VA copy object src.
+//! \param  [in] option
+//!         VA copy option, copy mode.
+//! \param  [in] sync_handle
+//!         VA copy sync handle
+//!
+//! \return VAStatus
+//!     VA_STATUS_SUCCESS if success, else fail reason
+//!
+VAStatus
+DdiMedia_Copy(
+    VADriverContextP    ctx,
+    VACopyObject       *dst_obj,
+    VACopyObject       *src_obj,
+    VACopyOption       option
+)
+{
+    VAStatus vaStatus = VA_STATUS_SUCCESS;
+    MOS_CONTEXT  mosCtx;
+    MOS_RESOURCE src, dst;
+    DdiCpInterface *pCpDdiInterface;
+
+    DDI_FUNCTION_ENTER();
+
+    DDI_CHK_NULL(ctx, "nullptr ctx", VA_STATUS_ERROR_INVALID_CONTEXT);
+    PDDI_MEDIA_CONTEXT mediaCtx   = DdiMedia_GetMediaContext(ctx);
+
+    DDI_CHK_NULL(mediaCtx,               "nullptr mediaCtx",               VA_STATUS_ERROR_INVALID_CONTEXT);
+    DDI_CHK_NULL(mediaCtx->pBufferHeap,  "nullptr mediaCtx->pBufferHeap",  VA_STATUS_ERROR_INVALID_CONTEXT);
+    DDI_CHK_NULL(mediaCtx->pSurfaceHeap, "nullptr mediaCtx->pSurfaceHeap", VA_STATUS_ERROR_INVALID_CONTEXT);
+    DDI_CHK_NULL(dst_obj, "nullptr copy dst", VA_STATUS_ERROR_INVALID_SURFACE);
+    DDI_CHK_NULL(src_obj, "nullptr copy src", VA_STATUS_ERROR_INVALID_SURFACE);
+
+    if (dst_obj->obj_type == VACopyObjectSurface)
+    {
+        DDI_CHK_LESS((uint32_t)dst_obj->object.surface_id, mediaCtx->pSurfaceHeap->uiAllocatedHeapElements, "copy_dst", VA_STATUS_ERROR_INVALID_SURFACE);
+        PDDI_MEDIA_SURFACE dst_surface = DdiMedia_GetSurfaceFromVASurfaceID(mediaCtx, dst_obj->object.surface_id);
+        DDI_CHK_NULL(dst_surface, "nullptr surface", VA_STATUS_ERROR_INVALID_SURFACE);
+        DDI_CHK_NULL(dst_surface->pGmmResourceInfo, "nullptr dst_surface->pGmmResourceInfo", VA_STATUS_ERROR_INVALID_PARAMETER);
+
+        MOS_ZeroMemory(&dst, sizeof(dst));
+        DdiMedia_MediaSurfaceToMosResource(dst_surface, &dst);
+    }
+    else if (dst_obj->obj_type == VACopyObjectBuffer)
+    {
+        DDI_CHK_LESS((uint32_t)dst_obj->object.buffer_id, mediaCtx->pBufferHeap->uiAllocatedHeapElements, "Invalid copy dst buf_id", VA_STATUS_ERROR_INVALID_BUFFER);
+        PDDI_MEDIA_BUFFER dst_buffer = DdiMedia_GetBufferFromVABufferID(mediaCtx, dst_obj->object.buffer_id);
+        DDI_CHK_NULL(dst_buffer, "nullptr buffer", VA_STATUS_ERROR_INVALID_BUFFER);
+        DDI_CHK_NULL(dst_buffer->pGmmResourceInfo, "nullptr dst_buffer->pGmmResourceInfo", VA_STATUS_ERROR_INVALID_PARAMETER);
+
+        MOS_ZeroMemory(&dst, sizeof(dst));
+        DdiMedia_MediaBufferToMosResource(dst_buffer, &dst);
+    }
```

Comment:
what about 'else' here?

Reply:
currently, buffer copy is not supported by vaCopy. buffer copy could be implemented by CPU. carl asked to keep buffer copy as further extension.

Reply:
If in VAAPI obj_type gets a new type, how the code above will work? There is no handling of "obj_type  != VACopyObjectSurface && obj_type  != VACopyObjectBuffer"

Reply:
if VAAPI update, suppose media-driver should be updated together. I think. the new type should be defined and verified and then be merged in.

Reply:
I was walking about error handling. Currently it's absent.

Reply:
okay, add it soon